### ISSUE TITLE
Module not loaded when it contains non-comment external statements

### ISSUE THREAD

**Issue by sccolbert** [systemjs/systemjs#629]
I've come across an issue where a statement above a `System.register` clause will cause the module not to load.

Distilled example:

`index.html`
```html
<!DOCTYPE html>
<html>
<head>
  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/systemjs/0.18.4/system.src.js"></script>
  <script type="text/javascript">
	System.config({ defaultJSExtensions: true });
	System.import('./foo').then(function(foo) {
	  console.log(foo);
	});
  </script>
</head>
<body>
</body>
</html>
```

`./foo.js`
```javascript
var a;  // this statement causes module not to load. comment-out and all works okay.
System.register([], function(exports) {
    console.log('load foo');
    exports("value", 42);
    return {
        setters:[],
        execute: function() {
            console.log('execute foo');
        }
    }
});
```

**Comment by guybedford** [comment]
Yes this is by design currently. We only detect modules as `System.register` that have that as the first line of code after comments and whitespace. This is so that the format detection can be performant. How did you generate this output, via TypeScript?

**Comment by sccolbert** [comment]
Yes, via TypeScript. Here's a real example of some generated output which occurs before the call to register:
```javascript
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};System.register(['../core/messaging', '../core/signaling', './nodewrapper'], function(exports_1) {...});
```

Also, a leading multi-line comment of the form `/* ... */` also causes the module loading to fail.

**Comment by guybedford** [comment]
@sccolbert thanks for the detailed update. Multi-line comments are definitely supported before a `System.register` statement - this is handled by the regex in https://github.com/systemjs/systemjs/blob/master/lib/register.js#L456.

If adding a custom `"use strict"` as part of post-processing then include a `"format register"` before it. Otherwise a leading `"use strict"` should be included inside the `System.register` wrapper during compilation as part of the format, which the helpers should be as well. This is being tracked in https://github.com/Microsoft/TypeScript/issues/3655.

**Comment by sccolbert** [comment]
@guybedford Thanks for the reply and link to TypeScript issue. Wrt to comments, it seem that regex doesn't handle a newline character in a multiline comment:
![rgx](https://cloud.githubusercontent.com/assets/137289/8984991/03fb0950-36a3-11e5-865b-0e5f9afa309f.PNG)

**Comment by sccolbert** [comment]
Also, I didn't come across anything in the docs about not having anything above a `System.register` call (but perhaps I missed it). I would classify this as a "gotcha" that would perhaps merit an FAQ entry.

**Comment by guybedford** [comment]
Thanks, I've fixed these in https://github.com/systemjs/systemjs/commit/7ab76a4b1e76f2c03e189336646f256fe4e25adb.

Yes it's a perf argument mainly - don't want to search entire source code when we can just search the first few lines carefully. System.register is the first module format detection, so that's why it's trying to do the least here. But it may turn out to be overkill as well.

Great point - I've added a new docs section at https://github.com/systemjs/systemjs/blob/master/docs/module-formats.md#module-format-detection.
